File: libs/core/src/binder/lookup_context.cpp
/**
 * @file lookup_context.cpp
 * @author Francesco Corso
 * @date 8/18/2026
 * @brief
 */
module prism.core:binder.lookup_context.impl;

import :binder.lookup_context;
import :diagnostics.diagnostic_bag;

namespace prism
{

    void LookupContext::report_diagnostic(Diagnostic diagnostic) const
    {
        if (diagnostics_ == nullptr)
            return;

        diagnostics_->add(std::move(diagnostic));
    }
} // namespace prism

File: src/Prism.Core/Binding/LookupContext.cs
using Prism.Core.Declarations;
using Prism.Core.Diagnostics;

namespace Prism.Core.Binding;

internal sealed class LookupContext
{
    private readonly DiagnosticBag? _diagnostics;

    private LookupContext(DiagnosticBag? diagnostics)
    {
        _diagnostics = diagnostics;
    }

    private static readonly LookupContext Empty = new(null);

    public static LookupContext Create(DiagnosticBag? diagnostics = null)
    {
        return diagnostics is not null ? new LookupContext(diagnostics) : Empty;
    }

    public void ReportDiagnostic(Diagnostic diagnostic)
    {
        _diagnostics?.Add(diagnostic);
    }
}
